core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -90,30 +93,239 @@
   /** Whether to cache blocks in this file into Alluxio. */
   private final boolean mShouldCache;
 
-  // The following 3 fields must be kept in sync. They are only updated in updateStreams together.
+  // The following variables should be kept in sync in updateStreams.
+  /** Position states. */
+  protected PositionState mPositionState;
+  /** Current block ID. */
+  private long mCurrentBlockId = UNINITIALIZED_BLOCK_ID;
+  /** Current block's start position. */
+  private long mCurrentBlockStartPos = UNINITIALIZED_BLOCK_POS;
+  /** Current block's end position (exclusive). */
+  private long mCurrentBlockEndPos = UNINITIALIZED_BLOCK_POS;
   /** Current block in stream backing this stream. */
   protected BlockInStream mCurrentBlockInStream;
   /**
    * Current block out stream writing the data into the local worker. This is only used when the in
    * stream reads from a remote worker.
    */
   protected BlockOutStream mCurrentCacheStream;
-  /** The blockId used in the block streams. */
-  private long mStreamBlockId;
 
   /** The read buffer in file seek. This is used in {@link #readCurrentBlockToEnd()}. */
   private final byte[] mSeekBuffer;
 
   /**
+   * States related to position that need to be in sync.
+   */
+  protected class PositionState {
+    /** Block ID used when position reaches EOF, should be < 0. */
+    public static final long EOF_BLOCK_ID = -1;
+
+    /** File length. */
+    private final long mFileLength;
```

Comment:
why fileLength belongs here? I thought this state object should be the state keeping changed

Reply:
I think it might need the file length to determine if it it at the end or not. It is the same reason that the block size is also in this class.

Reply:
Yes, Gene's understanding is correct.

Reply:
I was not clear. I mean: `mFileLength` is a constant, while my thought is that `PositionState`  keeps the moving parts related to position. So I feel `mFileLength` and `mBlockSize` is not part of a "state", but more like a constant w.r.t. this state, thus belong to the outer class. I am not very strong on this, just feel kinda wierd when reading code here. @gpang @uronce-cc